### PR TITLE
STCOM-451: Create date range filter

### PR DESCRIPTION
**Pull request by MaksymDryha**
This is a component for filtering records by a date range.
It has almost the same interface as CheckboxFilter and MultiSelectionFilter have. Unlike other filter components, this one does not require a `dataOptions` prop to be provided since the dates are manually entered by users and cannot be predefined. 
Another difference is that a `makeFilterString` function should be provided. When a user enters a date range, this function receives entered start and end dates and returns a string which will be used in the request parameters. Probably it can be given a better name.

Here is an example of using the component (inspired by ui-search):
```javascript
export default class Filters extends React.Component {
  static propTypes = {
    activeFilters: PropTypes.objectOf(PropTypes.array),
    onChange: PropTypes.func.isRequired,
  };

  onChangeHandler = (filter) => {
    this.props.onChange(filter);
  }

  retrieveDatesFromDateRangeFilterString(filterValue) {
    let dateRange = {
      startDate: '',
      endDate: '',
    };

    if (filterValue) {
      const [startDate, endDate] = filterValue[0].split(':');
      dateRange = {
        startDate: startDate || '',
        endDate: endDate || '',
      };
    }

    return dateRange;
  }

  makeDateRangeFilterString(startDate, endDate) {
    /**
    returned string will be passed to onChangeHandler and then used
    as the filter's value in the address bar.
    Given startDate is 1997-03-01 and endDate is 2000-01-21,
    then the address bar will look like:
    http://localhost:8080/codexsearch?filters=dateRange.1997-03-01%3A2000-01-21
    */
    return `${startDate}:${endDate}`; 
  }

  render() {
    const {
      activeFilters,
    } = this.props;

    return (
      <Accordion
        label="Date range"
        id="dateRange"
        name="dateRange"
      >
       <DateRangeFilter
          name="dateRange"
          selectedValues={this.retrieveDatesFromDateRangeFilterString(activeFilters.dateRange)}
          onChange={this.onChangeHandler}
          makeFilterString={this.makeDateRangeFilterString}
          dateFormat="YYYY-MM-DD"
        />
      </Accordion>
    );
  }
}
```

### CODE REVIEW

File: lib/SearchAndSort/components/DateRangeFilter/DateRangeFilter.js [hunk 1]
```diff
@@ -0,0 +1,285 @@
...
+    const {
+      dateFormat,
+      selectedValues: nextSelectedDates,
+    } = nextProps;
+
+    const { prevSelectedDates } = prevState;
+
+    let validationDataUpdates = {};
+
+    const areSelectedDatesSetFirstTime =
```
Comment (Godlevskyi) [inline lib/SearchAndSort/components/DateRangeFilter/DateRangeFilter.js]: Please cover cases with just one date recieved

File: lib/SearchAndSort/components/DateRangeFilter/DateRangeFilter.js [hunk 2]
```diff
@@ -0,0 +1,285 @@
...
+      endDate,
+    } = props.selectedValues;
+
+    this.startDateInput = React.createRef();
+    this.endDateInput = React.createRef();
+
+    const dateRangeIsNotEmpty = startDate !== '' || endDate !== '';
+    const validationData = dateRangeIsNotEmpty
+      ? validateDateRange(props.selectedValues)
+      : {};
```
Comment (Godlevskyi) [inline lib/SearchAndSort/components/DateRangeFilter/DateRangeFilter.js]: move default state

File: lib/SearchAndSort/components/DateRangeFilter/DateRangeFilter.js [hunk 3]
```diff
@@ -0,0 +1,285 @@
...
+
+    this.state = {
+      ...defaultFilterState,
```
Comment (Godlevskyi) [inline lib/SearchAndSort/components/DateRangeFilter/DateRangeFilter.js]: remove defaultFilterState

File: lib/SearchAndSort/components/DateRangeFilter/DateRangeFilter.js [hunk 4]
```diff
@@ -0,0 +1,285 @@
...
+      prevSelectedDates: {
+        startDate,
+        endDate,
+      },
+    };
+  }
+
+  onApplyButtonClick = () => {
+    const {
+      handleValidation: afterValidation,
```
Comment (Godlevskyi) [inline lib/SearchAndSort/components/DateRangeFilter/DateRangeFilter.js]: Please rename to applyFiltersIfValid

File: lib/SearchAndSort/components/DateRangeFilter/date-validations.js [hunk 5]
```diff
@@ -0,0 +1,42 @@
+import moment from 'moment';
+
+export const isDateValid = (date, dateFormat) => {
+  const momentDateWrapper = moment(date, dateFormat, true);
+
+  return momentDateWrapper.isValid();
+};
+
+export const validateDateRange = (dateRange, dateFormat) => {
```
Comment (zburke) [inline lib/SearchAndSort/components/DateRangeFilter/date-validations.js]: This is really nicely implemented. It's clear and easy to follow, and makes it easy to tell exactly which conditions failed. Nice work!

Comment (MaksymDryha) [reply lib/SearchAndSort/components/DateRangeFilter/date-validations.js]: Thank you @zburke!

**Pull request merged**